pingcap/tidb
# executor, sessionctx: calculate radix bit number before building hash table

PTAL @winoros @eurekaka 



### What problem does this PR solve? 
The first step of https://github.com/pingcap/tidb/issues/7762.

### What is changed and how it works?
1. This PR adds a new variable `L2CacheSize` in SessionVars to store the size of L2 cache.
2. This PR introduces a new package [github.com/klauspost/cpuid](https://github.com/klauspost/cpuid) to fetch the L2CacheSize.
2. Calculate the radix bit number after we read all the inner relation, the formula used is mentioned in proposal https://github.com/pingcap/tidb/blob/master/docs/design/2018-09-21-radix-hashjoin.md#proposal.
3. Note: We do not trigger the calculation for the radix bit number before the radix hash join is fully implemented. We use `UseRadixPartion` to control this according to the value of session variable "tidb_enable_radix_join". 

### Check List 

Tests 

 - Manual test
1. Set value of L2CacheSize to 1 in sessionctx/variable/session.go: NewSessionVars.
2. create table t(a bigint), and insert 8 values.
3. Add a log to check the value of HashJoinExec.radixBit which should be 8.
4. create table t(a bigint), and insert 64 values.
5. Add a log to check the value of HashJoinExec.radixBit which should be 9.

Code changes

 - Has exported variable/fields change

Side effects

none

Related changes

none
